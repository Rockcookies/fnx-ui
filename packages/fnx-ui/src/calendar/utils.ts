import { createBEM } from '../utils/namespace';

export const _bem = createBEM('fnx-calendar');

export type CalendarMode = 'single' | 'multiple' | 'range';

export interface CalendarMonthRef {
	root: HTMLDivElement | null;
	getHeight(): number;
}

export function compareMonth(d1: Date, d2: Date) {
	const year1 = d1.getFullYear();
	const year2 = d2.getFullYear();

	if (year1 === year2) {
		const month1 = d1.getMonth();
		const month2 = d2.getMonth();

		return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
	}

	return year1 > year2 ? 1 : -1;
}

export function compareDate(d1: Date, d2: Date) {
	const res = compareMonth(d1, d2);

	if (res === 0) {
		const date1 = d1.getDate();
		const date2 = d2.getDate();

		return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
	}

	return res;
}

export function isCalendarDateEquals(d1: Date, d2: Date) {
	return (
		d1.getFullYear() === d2.getFullYear() &&
		d1.getMonth() === d2.getMonth() &&
		d1.getDate() === d2.getDate()
	);
}

export const cloneDate = (date: Date) => new Date(date);

export const getPrevDay = (date: Date) => getDayByOffset(date, -1);
export const getNextDay = (date: Date) => getDayByOffset(date, 1);
export const getToday = () => {
	const today = new Date();
	today.setHours(0, 0, 0, 0);
	return today;
};

export function getDayByOffset(date: Date, offset: number) {
	const cloned = cloneDate(date);
	cloned.setDate(cloned.getDate() + offset);
	return cloned;
}

export function calcDateNum(d1: Date, d2: Date) {
	const day1 = d1.getTime();
	const day2 = d2.getTime();
	return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}
