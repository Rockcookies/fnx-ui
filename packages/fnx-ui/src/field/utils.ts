// native max length have incorrect line-break counting
// https://github.com/youzan/vant/issues/5033

import { InputHTMLAttributes } from 'react';
import { FieldInputType } from './interface';

// https://zhuanlan.zhihu.com/p/359265292
export function limitValueLength(value: string, maxLength?: number): string {
	const _value = [...value];
	if (maxLength != null && _value.length >= maxLength) {
		return _value.slice(0, maxLength).join('');
	}

	return value;
}

export const mapInputType = (
	inputType: FieldInputType,
): Partial<InputHTMLAttributes<any>> => {
	// type="number" is weird in iOS, and can't prevent dot in Android
	// so use inputmode to set keyboard in modern browsers
	if (inputType === 'number') {
		return {
			type: 'text',
			inputMode: 'decimal',
		};
	}

	if (inputType === 'digit') {
		return {
			type: 'tel',
			inputMode: 'numeric',
		};
	}

	return { type: inputType, inputMode: undefined };
};

/* export function getFieldId(
	...namePaths: Array<NamePath | null | undefined>
): string | undefined {
	const fid: string[] = [];

	for (const namePath of namePaths) {
		if (namePath != null) {
			const fields = (Array.isArray(namePath)
				? namePath
				: [namePath]
			).join('_');

			if (fields) {
				fid.push(fields);
			}
		}
	}

	if (fid.length > 0) {
		return fid.join('_');
	}
} */
