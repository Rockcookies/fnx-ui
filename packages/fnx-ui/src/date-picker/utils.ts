import { isDate } from '../utils/detect';

export const formatValue = (
	val: string | null | undefined,
): string | undefined => {
	if (val != null) {
		const v = val.trim();

		if (v) {
			return v;
		}
	}
};

export const iterateNumber = <T = string>(
	n: number,
	iteratee: (index: number) => T,
): T[] => {
	let index = -1;
	const result = Array<T>(n);

	while (++index < n) {
		result[index] = iteratee(index);
	}

	return result;
};

export class DateTime {
	private year = 0;
	private month = 0;
	private day = 0;
	private hour = 0;
	private minute = 0;
	private invalid = false;

	constructor(_date: Date | string | number) {
		const date = new Date(_date);

		const timestamp = date.getTime();

		if (isNaN(timestamp)) {
			this.invalid = false;
		} else {
			this.year = date.getFullYear();
			this.month = date.getMonth();
			this.day = date.getDate();
			this.hour = date.getHours();
			this.minute = date.getMinutes();
		}
	}

	isValid() {
		return !this.invalid;
	}

	getYear(): number {
		return this.year;
	}

	getMonth(): number {
		return this.month;
	}

	getDay(): number {
		return this.day;
	}

	getHour(): number {
		return this.hour;
	}

	getMinute(): number {
		return this.minute;
	}

	getDate(): Date {
		return new Date(
			this.year,
			this.month,
			this.day,
			this.hour,
			this.minute,
		);
	}

	getTime(): number {
		return this.getDate().getTime();
	}

	equals(dt: DateTime | Date) {
		if (isDate(dt)) {
			return (
				this.year === dt.getFullYear() &&
				this.month === dt.getMonth() &&
				this.day === dt.getDate() &&
				this.hour === dt.getHours() &&
				this.minute === dt.getMinutes()
			);
		}

		return (
			this.year === dt.getYear() &&
			this.month === dt.getMonth() &&
			this.day === dt.getDay() &&
			this.hour === dt.getHour() &&
			this.minute === dt.getMinute()
		);
	}
}
