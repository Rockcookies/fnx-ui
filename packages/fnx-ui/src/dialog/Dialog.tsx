import React, { forwardRef } from 'react';
import Button, { ButtonProps } from '../button';
import configComponentProps from '../hooks/config-component-props';
import { useLocale } from '../locale';
import Popup from '../popup';
import { BORDER_LEFT, BORDER_TOP } from '../utils/constants';
import { addUnit } from '../utils/format';
import { noop } from '../utils/misc';
import { classnames, createBEM } from '../utils/namespace';
import { DialogProps } from './interface';

const NS = 'fnx-dialog';
const bem = createBEM(NS);

const useProps = configComponentProps<
	Required<
		Pick<
			DialogProps,
			| 'messageAlign'
			| 'onConfirm'
			| 'onCancel'
			| 'confirmLoading'
			| 'confirmButtonProps'
			| 'cancelLoading'
			| 'cancelButtonProps'
			| 'transitionName'
			| 'bodyProps'
		>
	>
>({
	messageAlign: 'center',
	onConfirm: noop,
	onCancel: noop,
	confirmLoading: false,
	confirmButtonProps: {},
	cancelLoading: false,
	cancelButtonProps: {},
	transitionName: 'fnx-dialog-bounce',
	bodyProps: {},
});

const Dialog = forwardRef<HTMLDivElement, DialogProps>((_props, ref) => {
	const locale = useLocale('dialog');

	const [
		{
			messageAlign,
			onConfirm,
			onCancel,
			confirmLoading,
			cancelLoading,
			confirmButtonProps,
			cancelButtonProps,
			transitionName,
			bodyProps,
		},
		{
			title,
			footer,
			width,
			message,
			confirmButton,
			confirmText,
			cancelButton,
			cancelText,
			style,
			className,
			children,
			...resetProps
		},
	] = useProps(_props);

	const hasHeader = !!title;
	const hasBody = !!message || !!children;
	const hasFooter = footer !== false;

	const renderHeader = () => {
		return (
			<div
				className={bem('header', {
					isolated: !hasBody,
				})}
			>
				{title}
			</div>
		);
	};

	const renderBody = () => {
		if (children) {
			return (
				<div
					{...bodyProps}
					className={classnames(bem('body'), bodyProps.className)}
				>
					{children}
				</div>
			);
		}

		if (message) {
			return (
				<div
					className={bem('message', {
						'has-header': hasHeader,
						[messageAlign]: messageAlign,
					})}
				>
					{message}
				</div>
			);
		}
	};

	const renderButton = ({
		loading,
		children,
		onClick,
		className,
		extraProps,
	}: ButtonProps & { extraProps: ButtonProps }) => {
		return (
			<Button
				loading={loading}
				{...extraProps}
				className={classnames(className, extraProps.className)}
				onClick={(...args) => {
					onClick && onClick(...args);
					extraProps.onClick && extraProps.onClick(...args);
				}}
			>
				{children}
			</Button>
		);
	};

	const renderFooter = () => {
		if (footer) {
			return (
				<div className={classnames(BORDER_TOP, bem('footer'))}>
					{footer}
				</div>
			);
		}

		return (
			<div className={classnames(BORDER_TOP, bem('footer'))}>
				{confirmButton !== false &&
					(confirmButton ||
						renderButton({
							loading: confirmLoading,
							children: confirmText || locale.confirm,
							className: bem('confirm'),
							extraProps: confirmButtonProps,
							onClick: onConfirm,
						}))}
				{cancelButton !== false &&
					(cancelButton ||
						renderButton({
							loading: cancelLoading,
							children: cancelText || locale.cancel,
							className: classnames(
								confirmButton !== false
									? BORDER_LEFT
									: undefined,
								bem('cancel'),
							),
							extraProps: cancelButtonProps,
							onClick: onCancel,
						}))}
			</div>
		);
	};

	return (
		<Popup
			role="dialog"
			className={classnames(bem(), className)}
			style={{ width: addUnit(width), ...style }}
			ref={ref}
			overlayCloseable={false}
			transitionName={transitionName}
			{...resetProps}
			onClose={onCancel}
		>
			{hasHeader && renderHeader()}
			{hasBody && renderBody()}
			{hasFooter && renderFooter()}
		</Popup>
	);
});

Dialog.displayName = 'Dialog';

export default Dialog;
