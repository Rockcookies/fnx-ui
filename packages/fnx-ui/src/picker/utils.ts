import { clamp } from '../utils/format';

export function adjustActiveIndex<T>(
	index: number,
	options: T[],
	disabled: (o: T) => boolean,
): number {
	index = clamp(index, 0, Math.max(options.length - 1, 0));

	for (let i = index; i < options.length; i++) {
		if (!disabled(options[i])) return i;
	}
	for (let i = index - 1; i >= 0; i--) {
		if (!disabled(options[i])) return i;
	}

	return index;
}
