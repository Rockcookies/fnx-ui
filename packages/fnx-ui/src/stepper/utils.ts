import { formatNumber } from '../utils/format';
import { StepperValue } from './interface';

export function addValue(num1: number, num2: number) {
	const cardinal = 10 ** 10;
	return Math.round((num1 + num2) * cardinal) / cardinal;
}

export const getValidNumber = (
	v: StepperValue | undefined | null,
): number | undefined => {
	if (v == null) {
		return;
	}

	let valid: number | undefined;

	if (typeof v !== 'number') {
		valid = Number(
			formatNumber(String(v).trim(), {
				allowDot: true,
				allowMinus: true,
			}),
		);
	} else {
		valid = v;
	}

	if (valid == null || isNaN(valid)) {
		return;
	}

	return valid;
};
