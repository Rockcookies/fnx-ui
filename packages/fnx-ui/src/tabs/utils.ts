import { createBEM } from '../utils/namespace';
import { cancelRaf, raf } from '../utils/raf';

const NS = 'fnx-tabs';
export const _bem = createBEM(NS);

export const scrollLeftTo = (
	scrollElement: HTMLElement,
	to: number,
	duration: number,
): (() => void) => {
	let count = 0;
	const from = scrollElement.scrollLeft;
	const frames = duration === 0 ? 1 : Math.round(duration / 16);
	let rafId: number | undefined;

	function animate() {
		scrollElement.scrollLeft += (to - from) / frames;

		if (++count < frames) {
			rafId = raf(animate);
		} else {
			rafId = undefined;
		}
	}

	animate();

	return () => {
		if (rafId != null) {
			cancelRaf(rafId);
		}
	};
};
