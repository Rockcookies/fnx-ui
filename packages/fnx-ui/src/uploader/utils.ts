import { devWarning } from '../utils/misc';
import { createBEM } from '../utils/namespace';
import { UploaderFile } from './interface';

export const _bem = createBEM('fnx-uploader');

const now = +new Date();
let index = 0;

export function getFileUid() {
	return `fn-upload-${now}-${++index}`;
}

// https://github.com/react-component/upload/blob/master/src/attr-accept.ts
/* istanbul ignore next */
export function fileAccept(
	file: File,
	acceptedFiles?: string | string[],
): boolean {
	if (file && acceptedFiles) {
		const acceptedFilesArray = Array.isArray(acceptedFiles)
			? acceptedFiles
			: acceptedFiles.split(',');
		const fileName = file.name || '';
		const mimeType = file.type || '';
		const baseMimeType = mimeType.replace(/\/.*$/, '');

		return acceptedFilesArray.some((type) => {
			const validType = type.trim();
			// This is something like */*,*  allow all files
			if (/^\*(\/\*)?$/.test(type)) {
				return true;
			}

			// like .jpg, .png
			if (validType.charAt(0) === '.') {
				const lowerFileName = fileName.toLowerCase();
				const lowerType = validType.toLowerCase();

				let affixList = [lowerType];
				if (lowerType === '.jpg' || lowerType === '.jpeg') {
					affixList = ['.jpg', 'jpeg'];
				}

				return affixList.some((affix) => lowerFileName.endsWith(affix));
			}

			// This is something like a image/* mime type
			if (/\/\*$/.test(validType)) {
				return baseMimeType === validType.replace(/\/.*$/, '');
			}

			// Full match
			if (mimeType === validType) {
				return true;
			}

			// Invalidate type should skip
			if (/^\w+$/.test(validType)) {
				devWarning(
					false,
					'Uploader',
					`Upload takes an invalidate 'accept' type '${validType}'.Skip for check.`,
				);
				return true;
			}

			return false;
		});
	}
	return true;
}

const isImageFileType = (type: string): boolean => type.indexOf('image/') === 0;

export function readFileContent({
	file,
	content,
}: UploaderFile): Promise<string | undefined> {
	if (content != null) {
		return Promise.resolve(content || undefined);
	}

	if (file == null || file.type == null || !isImageFileType(file.type)) {
		return Promise.resolve(undefined);
	}

	return new Promise<string | undefined>((resolve) => {
		const reader = new FileReader();

		reader.onload = (event) => {
			resolve((event.target as FileReader).result as string);
		};

		reader.readAsDataURL(file);
	});
}
