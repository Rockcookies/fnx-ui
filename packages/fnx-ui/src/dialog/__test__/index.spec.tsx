import { act, fireEvent, render } from '@testing-library/react';
import React, { FC } from 'react';
import Dialog from '../';
import TestsDOM from '../../../test/dom';
import ConfigProvider from '../../config-provider';
import zhCN from '../../locale/zh-CN';
import { sleep } from '../../utils/misc';

describe('<Dialog/>', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.useRealTimers();
	});

	it('title、width、message、messageAlign props', () => {
		const { container } = render(
			<Dialog
				title="Dialog Title"
				width="80"
				message="content"
				messageAlign="right"
				visible={true}
			></Dialog>,
		);

		expect(container).toMatchSnapshot();
	});

	it('confirmButton、confirmText、confirmButtonProps props', () => {
		const { rerender } = render(
			<Dialog
				visible={true}
				confirmText="Dialog Confirm"
				confirmButtonProps={{ disabled: true }}
			></Dialog>,
		);

		const confirm = TestsDOM.mustQuerySelector(
			document.body,
			'.fnx-button--disabled',
		);

		expect(confirm).not.toBeNull();
		expect(
			TestsDOM.mustQuerySelector(confirm, '.fnx-button__text').innerHTML,
		).toEqual('Dialog Confirm');

		rerender(
			<Dialog
				visible={true}
				confirmButton={<span className="define-btn">confirm</span>}
			></Dialog>,
		);

		expect(
			TestsDOM.mustQuerySelector(document.body, '.define-btn'),
		).not.toBeNull();
	});

	it('cancelButton、cancelText、cancelButtonProps props', () => {
		const { rerender } = render(
			<Dialog
				visible
				cancelText="Dialog Cancel"
				cancelButtonProps={{ className: 'cancel-button' }}
			></Dialog>,
		);

		const cancel = TestsDOM.mustQuerySelector(
			document.body,
			'.cancel-button',
		);

		expect(cancel).not.toBeNull();
		expect(
			TestsDOM.mustQuerySelector(cancel, '.fnx-button__text').innerHTML,
		).toEqual('Dialog Cancel');

		rerender(
			<Dialog
				visible={true}
				confirmButton={<span className="define-btn">confirm</span>}
			></Dialog>,
		);

		expect(
			TestsDOM.mustQuerySelector(document.body, '.define-btn'),
		).not.toBeNull();
	});

	it('footer props', () => {
		render(
			<Dialog
				visible
				footer={<div className="dialog-footer"></div>}
			></Dialog>,
		);

		const footer = TestsDOM.mustQuerySelector(
			document.body,
			'.dialog-footer',
		);

		expect(footer).not.toBeNull();
	});

	it('onConfirm、onCancel', async () => {
		const onConfirm = jest.fn(() => {
			return sleep(2000);
		});
		const onCancel = jest.fn(() => {
			return false;
		});

		act(() => {
			Dialog.show({
				title: 'title',
				message: 'content',
				onConfirm,
				onCancel,
				mountTo: document.body,
			});
		});

		const confirm = TestsDOM.mustQuerySelector(
			document.body,
			'.fnx-dialog__confirm',
		);
		const cancel = TestsDOM.mustQuerySelector(
			document.body,
			'.fnx-dialog__cancel',
		);

		act(() => {
			fireEvent.click(cancel);
		});
		expect(onCancel).toHaveBeenCalled();

		act(() => {
			fireEvent.click(confirm);
		});
		expect(onConfirm).toHaveBeenCalled();
		expect(confirm.classList.contains('fnx-button--loading')).toBeTruthy();
	});

	it('close、destroy function', async () => {
		act(() => {
			Dialog.show({
				title: 'title',
				message: 'content',
			});
		});

		act(() => {
			Dialog.show({
				title: 'title2',
				message: 'content2',
			});
		});

		act(() => {
			jest.runAllTimers();
		});

		const dialogDiv = TestsDOM.mustQuerySelector(
			document.body,
			'.fnx-dialog',
		);
		expect(dialogDiv).not.toBeNull();

		act(() => {
			Dialog.clearAll();
		});
		act(() => {
			jest.runAllTimers();
		});

		expect(document.body).toMatchSnapshot();
	});

	it('useDialog', async () => {
		const Demo: FC = () => {
			const { show } = Dialog.useDialog();

			return (
				<div
					className="trigger"
					onClick={() =>
						show({
							title: 'title',
							message: 'message',
						})
					}
				></div>
			);
		};

		const element = document.createElement('div');

		const { container } = render(
			<ConfigProvider mountTo={() => element} locale={zhCN}>
				<Demo />
			</ConfigProvider>,
		);

		const trigger = TestsDOM.mustQuerySelector(container, '.trigger');

		act(() => {
			fireEvent.click(trigger);
		});

		expect(element).toMatchSnapshot();
	});
});
