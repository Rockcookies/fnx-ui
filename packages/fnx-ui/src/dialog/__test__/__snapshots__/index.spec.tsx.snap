// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dialog/> title、width、message、messageAlign props 1`] = `<div />`;

exports[`<Dialog/> useDialog 1`] = `
<div>
  <div>
    <div
      class="fnx-overlay fnx-fade-enter fnx-fade-enter-active"
      style="z-index: 2008;"
    />
    <div
      class="fnx-popup fnx-popup--center fnx-dialog fnx-dialog-bounce-enter fnx-dialog-bounce-enter-active"
      role="dialog"
      style="z-index: 2008;"
    >
      <div
        class="fnx-dialog__header"
      >
        title
      </div>
      <div
        class="fnx-dialog__message fnx-dialog__message--has-header fnx-dialog__message--center"
      >
        message
      </div>
      <div
        class="fnx-hairline--top fnx-dialog__footer"
      >
        <button
          class="fnx-button fnx-button--default fnx-button--md fnx-dialog__confirm"
        >
          <span
            class="fnx-button__content"
          >
            <span
              class="fnx-button__text"
            >
              确定
            </span>
          </span>
        </button>
        <button
          class="fnx-button fnx-button--default fnx-button--md fnx-hairline--left fnx-dialog__cancel"
        >
          <span
            class="fnx-button__content"
          >
            <span
              class="fnx-button__text"
            >
              取消
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
